### PR TITLE
[New resources] IntuneTrustedRootCertificateAndroidDeviceOwner and IntuneTrustedRootCertificateAndroidEnterprise

### PR DESCRIPTION
**Pull request by dannyKBjj**
Initial release. 2 DSC resource to cover both Android Enterprise and Android Device Owner/Administrator certificate policies.

#### Pull Request (PR) description
2 DSCResource Modules to cover Android Trusted Root Cert policies for both Enterprise and Device Owner

#### This Pull Request (PR) fixes the following issues
None

#### Task list

- [x] Added an entry to the change log under the Unreleased section of the file CHANGELOG.md.
      Entry should say what was changed and how that affects users (if applicable), and
      reference the issue being resolved (if applicable).
- [x] Resource parameter descriptions added/updated in the schema.mof.
- [x] Resource documentation added/updated in README.md.
- [x] Resource settings.json file contains all required permissions.
- [x] Examples appropriately added/updated.
- [x] Unit tests added/updated.
- [x] New/changed code adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

### CODE REVIEW

File: Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner.schema.mof
Comment (ricmestre) [inline Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner.schema.mof:1]: Change to UTF8 and try again

File: Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateAndroidEnterprise/MSFT_IntuneTrustedRootCertificateAndroidEnterprise.schema.mof
Comment (ricmestre) [inline Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateAndroidEnterprise/MSFT_IntuneTrustedRootCertificateAndroidEnterprise.schema.mof:1]: Change to UTF8 and try again

File: Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner.schema.mof
Comment (dannyKBjj) [reply Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner.schema.mof:1]: Done

File: Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateAndroidEnterprise/MSFT_IntuneTrustedRootCertificateAndroidEnterprise.schema.mof
Comment (dannyKBjj) [reply Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateAndroidEnterprise/MSFT_IntuneTrustedRootCertificateAndroidEnterprise.schema.mof:1]: Done

**Comment by FabienTschanz** [comment]
@dannyKBjj Do you know that we have a [resource generator](https://microsoft365dsc.com/user-guide/advanced/dynamic-resource-generator/) available in M365DSC? It takes care of most of the stuff, including generating all of the properties, structure, correct naming and much more.

**Comment by ykuijs** [comment]
@ricmestre Can this PR be merged? Has it been fully reviewed?

**Comment by dannyKBjj** [comment]
> @dannyKBjj Do you know that we have a [resource generator](https://microsoft365dsc.com/user-guide/advanced/dynamic-resource-generator/) available in M365DSC? It takes care of most of the stuff, including generating all of the properties, structure, correct naming and much more.

No I wasn't aware of it. I've got an open PR for iOS VPN i need to work on today and I'll try that tool out for the next one.

File: Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner.schema.mof
Comment (ricmestre) [inline Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner.schema.mof:1]: This one still seems to be UTF16

File: Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner.psm1 [hunk 1]
```diff
@@ -0,0 +1,701 @@
...
+    Add-M365DSCTelemetryEvent -Data $data
+    #endregion   
+
+    $nullResult = $PSBoundParameters
+    $nullResult.Ensure = 'Absent'
+    try
+    {
+        if (-not [string]::IsNullOrWhiteSpace($id))
+        { 
+		    $getValue = Get-MgBetaDeviceManagementDeviceConfiguration -DeviceConfigurationId $id -ErrorAction SilentlyContinue
```
Comment (ricmestre) [inline Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner.psm1]: Please replace to 4 spaces

**Comment by ricmestre** [comment]
@ykuijs @dannyKBjj Both resources seems to work without issues, but I thought I was going crazy and had to check several times and they are correct but *IntuneTrustedRootCertificateAndroidDeviceOwner* is creating an *Android Enterprise* certificate and *IntuneTrustedRootCertificateAndroidEnterprise* creates an *Android device administrator*, do you also see this? At least the code seems to be correct and using the right corresponding *@odata.type*

Could you please just double check this for me that it's not my machine acting up?

**Comment by ricmestre** [comment]
Just see below, do you have the same issue? Is this a Graph problem?

image

**Comment by ricmestre** [comment]
OK, I'm not crazy. I ran the Graph cmdlets manually and they return correctly but the admin portal looks like they show up reversed, at least to me. Probably it's just the way it works? I have to be honest here that I'm not too familiar with how Android policies should display in the admin portal.

File: Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner.psm1 (hunk 1 above)
Comment (dannyKBjj) [reply Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner.psm1]: done

**Comment by dannyKBjj** [comment]
> @ykuijs @dannyKBjj Both resources seems to work without issues, but I thought I was going crazy and had to check several times and they are correct but _IntuneTrustedRootCertificateAndroidDeviceOwner_ is creating an _Android Enterprise_ certificate and _IntuneTrustedRootCertificateAndroidEnterprise_ creates an _Android device administrator_, do you also see this? At least the code seems to be correct and using the right corresponding _@odata.type_
> 
> Could you please just double check this for me that it's not my machine acting up?

Oh, yes you seem to be right. I must admit I tested both policies at once for export/create/update etc. and didn't notice. I have just tried exporting with IntuneTrustedRootCertificateAndroidEnterprise only and it has exported my 'DeviceOwner' policies!?

just replicated the problem calling New-MgBetaDeviceManagementDeviceConfiguration directly.

**Comment by dannyKBjj** [comment]
It isn't the MgBetaDeviceManagementDeviceConfiguration cmdlet either, i tried the same using Invoke-MgGraphRequest

Same result... it isn't the behaviour I'd expect. What do I do? Swap over the oData.Type ?! Currently the module names suggest they handle the opposite of what they do.

**Comment by ricmestre** [comment]
Thing is your modules are doing the right thing in Graph, it's just the way they are presented in the admin portal that seem to be reversed, I think I had the same "problem" with similar Enterprise/DeviceOwner policies.

@ykuijs Do you know anything about this?

File: Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner.schema.mof
Comment (dannyKBjj) [reply Modules/Microsoft365DSC/DSCResources/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner/MSFT_IntuneTrustedRootCertificateAndroidDeviceOwner.schema.mof:1]: Resaved everything as UTF-8

**Comment by ykuijs** [comment]
No, I am unfamiliar with this issue. Let me check with someone at the PG.

**Comment by ricmestre** [comment]
@ykuijs Thank you, just to let you know that everything works as expected code-wise so PR should still be merged regardless, I opened the developer tool in the browser and the requests also all look ok, it's just the way that the admin portal shows the "Platform" property reversed but like I said I think this also affects other Android policies (device compliance, configuration, etc) so please have them know to also look into those.

**Comment by ykuijs** [comment]
@ricmestre, what do you mean when you say this also affects other Android policies? Is the portal displaying the certificate incorrectly? Or are other policies also showed incorrectly? Can you share an example?

FYI: Response from the PG is a little slower because of Thanksgiving.

**Comment by ricmestre** [comment]
@ykuijs I can't test every Android related policies but see below, from those shown only 2 have the correct platform. Again the actual policies Graph wise are actually correct it's just the display in the portal that it's incorrect.

image

**Comment by ricmestre** [comment]
And to stir the pot a little bit more, you can only create these device restrictions through the portal, so M365DSC has one more for Work Profile which is only available via API and in the portal gets the "Enterprise" platform as shown in the previous screenshot.

image

**Comment by dannyKBjj** [comment]
Hi, what's the status with this PR please? Will it be merged?

**Comment by ykuijs** [comment]
I kept this PR open because I am still in contact with the PG. But am planning to merge it before tomorrows release.

**Comment by dannyKBjj** [comment]
I'm currently doing an Android VPN dsc resource and am finding the same issue:

microsoft.graph.androidVpnConfiguration seems to be for 'device Administrator/Owner' policies

microsoft.graph.androidDeviceOwnerVpnConfiguration seems to be for Enterprise policies

leaves me wondering whether my DSC Resource name should follow the 'resource type' name, or the reality of what it seems to correspond to within Intune?! For these trusted certificate modules, the name corresponds to the resource type, so I guess I should do the same for the VPN modules (assuming we do merge this PR?)

**Comment by ykuijs** [comment]
Please use resource type in this case. I am in contact with the PG to have this checked. Would be a waste of time if we have to revert this again when it is corrected in the portal

**Pull request merged**